Foundation/URLSession/http/HTTPURLProtocol.swift

```diff
@@ -189,16 +163,78 @@ fileprivate extension _HTTPURLProtocol {
             self?.completeTask(withError: urlError)
             self?.client?.urlProtocol(self!, didFailWithError: urlError)
         }
-	guard let task = self.task else { fatalError() }
+        guard let task = self.task else { fatalError() }
         easyHandle.timeoutTimer = _TimeoutSource(queue: task.workQueue, milliseconds: timeoutInterval, handler: timeoutHandler)
-
+        
         easyHandle.set(automaticBodyDecompression: true)
         easyHandle.set(requestMethod: request.httpMethod ?? "GET")
         if request.httpMethod == "HEAD" {
             easyHandle.set(noBody: true)
         }
     }
+    
+    /// What action to take
+    override func completionAction(forCompletedRequest request: URLRequest, response: URLResponse) -> _CompletionAction {
+        // Redirect:
+        let httpURLResponse = response as!  HTTPURLResponse
+        if let request = redirectRequest(for: httpURLResponse, fromRequest: request) {
+            return .redirectWithRequest(request)
+        }
+        return .completeTask
+    }
+    
+    override func redirectFor(request: URLRequest) {
+        //TODO: Should keep track of the number of redirects that this
+        // request has gone through and err out once it's too large, i.e.
+        // call into `failWith(errorCode: )` with NSURLErrorHTTPTooManyRedirects
+        guard case .transferCompleted(response: let response, bodyDataDrain: let bodyDataDrain) = self.internalState else {
```

Comment:
`response` and `bodyDataDrain` don't seem to be used?

Reply:
Its been used  .